draft-ietf-quic-tls.md

```diff
@@ -255,142 +205,203 @@ A simplified TLS 1.3 handshake with 0-RTT application data is shown in
-The initial packet from a client contains a stream frame for stream 0 that
-contains the first TLS handshake messages from the client.  This allows the TLS
-handshake to start with the first packet that a client sends.
+~~~
++------------+                        +------------+
+|            |<- Handshake Messages ->|            |
+|            |<---- 0-RTT Keys -------|            |
+|            |<--- Handshake Keys-----|            |
+|   QUIC     |<---- 1-RTT Keys -------|    TLS     |
+|            |<--- Handshake Done ----|            |
++------------+                        +------------+
+ |         ^
+ | Protect | Protected
+ v         | Packet
++------------+
+|   QUIC     |
+|  Packet    |
+| Protection |
++------------+
+~~~
+{: #schematic title="QUIC and TLS Interactions"}
 
-QUIC packets are protected using a scheme that is specific to QUIC, see
-{{packet-protection}}.  Keys are exported from the TLS connection when they
-become available using a TLS exporter (see Section 7.5 of {{!TLS13}} and
-{{key-expansion}}).  After keys are exported from TLS, QUIC manages its own key
-schedule.
+Unlike TLS over TCP, QUIC applications which want to send data do not
+send it through TLS "application_data" records. Rather, they send it
+as QUIC STREAM frames which are then carried in QUIC packets.
+
+
+# Carrying TLS Messages {#carrying-tls}
+
+QUIC carries TLS handshake data in CRYPTO_HS frames, each of which
+consists of a contiguous block of handshake data (identified by an
+offset and length). Those frames are packaged into QUIC packets
+and encrypted under the current TLS encryption level.
+As with TLS over TCP, once TLS handshake data has
+been delivered to QUIC, it is QUIC's responsibility to deliver it
+reliably. Each chunk of data is associated with the then-current TLS
+sending keys, and if QUIC needs to retransmit that data, it MUST use
+the same keys even if TLS has already updated to newer keys.
+
+One important difference between TLS 1.3 records (used with TCP)
+and QUIC CRYPTO_HS frames is that in QUIC multiple frames may appear
+in the same QUIC packet as long as they are associated with the
+same encryption level. For instance, an implementation might
+bundle a Handshake message and an ACK for some Handshake
+data into the same packet.
+
+In general, the rules for which data can appear in packets of which
+encryption level are the same in QUIC as in TLS over TCP:
+
+- CRYPTO frames MAY appear in packets of any encryption level.
+- CONNECTION_CLOSE and CRYPTO_CLOSE MAY appear in packets of any
+  encryption level other than 0-RTT.
+- PADDING frames MAY appear in packets of any encryption level.
+- ACK frames MAY appear in packets of any encryption level, but
+  MUST only acknowledge packets which appeared in that encryption
+  level.
+- STREAM frames MAY appear in the 0-RTT and 1-RTT levels.
+- All other frame types MUST only appear at the 1-RTT levels.
```

Comment:
PING frames should be allowed at any encryption level. They are really useful for implementing TLPs (just send a packet containing a single PING frame, if there's no new data to send).

Reply:
Done